### PR TITLE
Record NuGet PackageDownloads

### PR DESCRIPTION
**Pull request by ericstj**
Read PackageDownloads from the assets file and record them as top-level development dependencies.

### CODE REVIEW

**Comment by ericstj** [comment]
@grvillic -- thoughts on this one?

File: src/Microsoft.ComponentDetection.Detectors/nuget/NuGetPackageReferenceFrameworkAwareDetector.cs [hunk 1]
```diff
@@ -141,16 +144,35 @@ private void NavigateAndRegister(
...
     }
 
-    private List<(string Name, Version Version, VersionRange VersionRange)> GetTopLevelLibraries(LockFile lockFile)
+    private void RegisterPackageDownloads(ISingleFileComponentRecorder singleFileComponentRecorder, LockFile lockFile)
+    {
+        foreach (var framework in lockFile.PackageSpec.TargetFrameworks)
+        {
+            foreach (var packageDownload in framework.DownloadDependencies)
+            {
+                var libraryComponent = new DetectedComponent(new NuGetComponent(packageDownload.Name, packageDownload.VersionRange.MinVersion.ToNormalizedString()));
```
Comment (grvillic) [inline src/Microsoft.ComponentDetection.Detectors/nuget/NuGetPackageReferenceFrameworkAwareDetector.cs:153]: I don't know if this is even possible, but let's skip Download dependencies if package download name or version range are null or empty. We can log an informational message when that is the case. Just safeguarding any invalid manifest files that are missing metadata or are corrupted.

File: src/Microsoft.ComponentDetection.Detectors/nuget/NuGetPackageReferenceFrameworkAwareDetector.cs [hunk 2]
```diff
@@ -141,16 +144,35 @@ private void NavigateAndRegister(
...
+                    targetFramework: framework.FrameworkName?.GetShortFolderName());
+            }
+        }
+    }
+
+    private List<(string Name, Version Version, VersionRange VersionRange, bool IsPackageDownload)> GetTopLevelLibraries(LockFile lockFile)
     {
         // First, populate libraries from the TargetFrameworks section -- This is the base level authoritative list of nuget packages a project has dependencies on.
-        var toBeFilled = new List<(string Name, Version Version, VersionRange VersionRange)>();
+        var toBeFilled = new List<(string Name, Version Version, VersionRange VersionRange, bool IsPackageDownload)>();
```
Comment (grvillic) [inline src/Microsoft.ComponentDetection.Detectors/nuget/NuGetPackageReferenceFrameworkAwareDetector.cs:169]: Are we using this new property `IsPackageDownload`? Doesn't seem like it.

**Review by grvillic — CHANGES_REQUESTED** [review]
Minor comments

File: src/Microsoft.ComponentDetection.Detectors/nuget/NuGetPackageReferenceFrameworkAwareDetector.cs (hunk 1 above)
Comment (ericstj) [reply src/Microsoft.ComponentDetection.Detectors/nuget/NuGetPackageReferenceFrameworkAwareDetector.cs:153]: I don't ever expect these to be null or have null values so we can just skip them.  https://github.com/NuGet/NuGet.Client/blob/38729596a6549df3cd1eed37b93d51df1bf62bff/src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.Utf8JsonStreamReader.cs#L746

**Review by grvillic — APPROVED** [review]

**Pull request merged**